ack_timer.go

```diff
@@ -41,34 +52,35 @@ func (t *ackTimer) start() bool {
 	t.mutex.Lock()
 	defer t.mutex.Unlock()
 
-	// this timer is already closed
-	if t.closed {
+	// this timer is already closed or already running
+	if t.state != ackTimerStopped {
 		return false
 	}
-
-	// this is a noop if the timer is already running
-	if t.stopFunc != nil {
-		return false
-	}
-
-	cancelCh := make(chan struct{})
+	t.state = ackTimerStarted
 
 	go func() {
-		timer := time.NewTimer(t.interval)
+		t.timer.Reset(ackInterval)
 
 		select {
-		case <-timer.C:
-			t.stop()
-			t.observer.onAckTimeout()
-		case <-cancelCh:
-			timer.Stop()
+		case <-t.timer.C:
+			t.mutex.Lock()
+			switch t.state {
+			case ackTimerStopped:
+				// if the timer is already stopped empty the cancel channel
+				<-t.cancel
+			case ackTimerStarted:
+				t.state = ackTimerStopped
+				defer t.observer.onAckTimeout()
+			case ackTimerClosed:
```

Comment:
is it worth emptying t.cancel here as well?

Reply:
state and writes to the cancel channel are synced. 
* if we are in this critical section and the state is `started` we cannot have written to `cancel`. 
* if we change the state from `started` to `stopped` in this critical section then call `stop` afterward the state will be `stopped` and we won't won't write to `cancel`.
* `cancel` could be dirty if the state is `closed` but we won't reuse the timer later so we don't care.